### PR TITLE
[#10264] Remove Instructor Google ID from Output

### PR DESCRIPTION
**Pull request by madanalogy**
Part of #10264 

Only the instructor can see his own Google ID and the owner/co-owners of the course that the instructor is part of.

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/action/GetInstructorsAction.java [hunk 1]
```diff
@@ -76,7 +77,18 @@ public ActionResult execute() {
             });
         } else if (intentStr.equals(Intent.FULL_DETAIL.toString())) {
             // get all instructors of a course without information hiding
-            data = new InstructorsData(instructorsOfCourse);
+            // adds googleId if caller is course co-owner or admin
+            if (userInfo.isAdmin || logic.getInstructorForGoogleId(courseId, userInfo.getId()).getRole().equals(
+                    Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_COOWNER)) {
```
Comment (madanalogy) [inline src/main/java/teammates/ui/webapi/action/GetInstructorsAction.java:82]: After much debugging I've isolated the problem with the e2e tests to the addition of this line. Whenever `verifyPresentInDatastore` is called in a test case for instructors, the `BackDoor` will attempt to execute a `GET` request on the `/instructors` endpoint. This will now however always throw a NPE because `userInfo.getId()` is null. Any advice to help resolve this would be appreciated @wkurniawan07 @xpdavid @jtankw3

Comment (madanalogy) [reply src/main/java/teammates/ui/webapi/action/GetInstructorsAction.java:82]: I was hoping against that because the `getMinAuthLevel` is already defined as `LOGGED_IN`. There appears to be another issue as well with `UpdateInstructorAction`. I'm working with @jtankw3 to help resolve it

Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/action/GetInstructorsAction.java:82]: By design, the `execute()` method should not care about access control (this is the complete seperation of access control we want to achieve in V7) so we are not breaking any promise.

File: src/main/java/teammates/ui/webapi/action/GetInstructorAction.java [hunk 2]
```diff
@@ -51,24 +51,32 @@ public ActionResult execute() {
...
             break;
         default:
             throw new InvalidHttpParameterException("Unknown intent " + intent);
         }
 
-        if (instructorAttributes == null) {
-            return new JsonResult("Instructor could not be found for this course", HttpStatus.SC_NOT_FOUND);
-        }
-
-        return new JsonResult(new InstructorData(instructorAttributes));
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/GetInstructorAction.java:71]: The previous implementation is DRY, this one is not. Should only add some line of code to populate Google ID when necessary, e.g.
```java
if (instructorAttributes == null) {
    return new JsonResult("Instructor could not be found for this course", HttpStatus.SC_NOT_FOUND);
}

InstructorData instructorData = new InstructorData(instructorAttributes);
if (intent == Intent.FULL_DETAIL) {
    instructorData.setGoogleId(instructorAttributes.googleId);
}

return new JsonResult(instructorData);
```

File: src/main/java/teammates/ui/webapi/action/UpdateInstructorAction.java [hunk 3]
```diff
@@ -69,7 +69,9 @@ public ActionResult execute() {
                                 .withRole(instructorToEdit.role)
                                 .build());
             }
-            return new JsonResult(new InstructorData(updatedInstructor));
+            InstructorData newInstructorData = new InstructorData(updatedInstructor);
+            newInstructorData.setGoogleId(updatedInstructor.getGoogleId());
+            return new JsonResult(newInstructorData);
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/UpdateInstructorAction.java:74]: This might actually count as Google ID leaking, but I guess someone who has the privilege of updating instructor has no harm of knowing the Google ID.

File: src/test/java/teammates/test/cases/webapi/SearchInstructorsActionTest.java [hunk 4]
```diff
@@ -52,9 +52,7 @@ protected void testExecute_searchCourseId_shouldSucceed() {
         JsonResult result = getJsonResult(action);
         InstructorsData response = (InstructorsData) result.getOutput();
         assertTrue(response.getInstructors().stream()
-                .filter(i -> i.getName().equals(acc.getName()))
-                .findAny()
-                .isPresent());
+                .anyMatch(i -> i.getName().equals(acc.getName())));
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/webapi/SearchInstructorsActionTest.java:55]: All the changes in this file are unrelated

File: src/e2e/java/teammates/e2e/util/BackDoor.java [hunk 5]
```diff
@@ -333,9 +333,10 @@ public static CourseAttributes getArchivedCourse(String instructorId, String cou
     /**
      * Gets instructor data from the datastore.
      */
-    public static InstructorData getInstructorData(String courseId, String email) {
+    public static InstructorData getInstructorData(String courseId, String instructorId, String email) {
         Map<String, String[]> params = new HashMap<>();
         params.put(Const.ParamsNames.COURSE_ID, new String[] { courseId });
+        params.put(Const.ParamsNames.USER_ID, new String[] { instructorId });
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/util/BackDoor.java:339]: Do we still need this `USER_ID` parameter, now that you are guaranteed to be admin if using the backdoor method?

File: src/main/java/teammates/ui/webapi/action/Action.java [hunk 6]
```diff
@@ -98,6 +98,9 @@ private void initAuthInfo() {
         if (Config.BACKDOOR_KEY.equals(req.getHeader("Backdoor-Key"))) {
             authType = AuthType.ALL_ACCESS;
             userInfo = new UserInfo(getRequestParamValue(Const.ParamsNames.USER_ID));
+            userInfo.isAdmin = true;
+            userInfo.isStudent = true;
+            userInfo.isInstructor = true;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/Action.java:103]: We might need to be able to customize this, but it is for another time.

File: src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts [hunk 7]
```diff
@@ -285,7 +285,7 @@ export class InstructorCourseEditPageComponent implements OnInit {
      * typed as such to accomodate for a use case in SearchService.
      */
     return {
-      googleId: i.googleId,
+      googleId: i.googleId ? i.googleId : '',
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts:288]: Check the front-end also. If the `googleId` is not available, the whole row should not be shown (instead of showing empty input box).

File: src/test/java/teammates/test/cases/webapi/GetInstructorActionTest.java [hunk 8]
```diff
@@ -59,6 +59,7 @@ protected void testExecute() throws Exception {
         assertEquals(HttpStatus.SC_OK, actionOutput.getStatusCode());
         InstructorData response = (InstructorData) actionOutput.getOutput();
         assertEquals(instructor1OfCourse1.name, response.getName());
+        assertNull(response.getGoogleId());
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/webapi/GetInstructorActionTest.java:62]: Registration key should be asserted as null also

File: src/main/java/teammates/ui/webapi/action/SearchInstructorsAction.java [hunk 9]
```diff
@@ -49,7 +49,8 @@ public ActionResult execute() {
             InstructorData instructorData = new InstructorData(instructor);
             instructorData.addAdditionalInformationForAdminSearch(
                     StringHelper.encrypt(instructor.getKey()),
-                    getInstituteFromGoogleId(instructor.getGoogleId()));
+                    getInstituteFromGoogleId(instructor.getGoogleId()),
+                    instructor.getGoogleId());
             instructorData.hideInformationForSearch();
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/SearchInstructorsAction.java:54]: Only admins can search instructors, and we have no reason to hide anything from admin. This line can be removed.

File: src/web/app/pages-instructor/instructor-course-edit-page/instructor-edit-panel/instructor-edit-panel.component.ts [hunk 10]
```diff
@@ -8,7 +8,7 @@ import {
  * Model for edit instructor panel.
  */
 export interface InstructorEditPanel {
-  googleId: string;
+  googleId?: string;
```
Comment (wkurniawan07) [inline src/web/app/pages-instructor/instructor-course-edit-page/instructor-edit-panel/instructor-edit-panel.component.ts:11]: https://github.com/TEAMMATES/teammates/blob/1df7a07640bb533ab4202a2b745107e2e64cfd5e/src/web/app/pages-instructor/instructor-course-edit-page/instructor-edit-panel/instructor-edit-panel.component.html#L40-L50

Need to modify the front-end slightly. This row should be visible only if the user has `canmodifyinstructor` permission.

File: src/main/java/teammates/ui/webapi/action/GetInstructorsAction.java (hunk 1 above)
Comment (wkurniawan07) [inline src/main/java/teammates/ui/webapi/action/GetInstructorsAction.java:82]: I have checked the access control logic again. It is not necessarily only co-owner can see the Google ID, but rather anyone who has `canmodifyinstructor` permission.
